src/MagnumPlugins/UfbxImporter/UfbxImporter.cpp

```diff
@@ -0,0 +1,1162 @@
+            for (UnsignedInt set = 0; set < bitangentSetCount; ++set)
+                bitangentSets[set][dstIx] = Vector3(mesh->uv_sets[set].vertex_bitangent[srcIx]);
+            for (UnsignedInt set = 0; set < colorSetCount; ++set)
+                colorSets[set][dstIx] = Color4(mesh->color_sets[set].vertex_color[srcIx]);
+            dstIx++;
+        }
+    }
+
+    Containers::Array<char> indexData{NoInit, indexCount*sizeof(UnsignedInt)};
+    Containers::ArrayView<UnsignedInt> indices = Containers::arrayCast<UnsignedInt>(indexData);
+
+    /* The vertex data is unindexed, so generate a contiguous index range */
+    for (UnsignedInt i = 0; i < indexCount; i++)
+        indices[i] = i;
+
+    MeshData meshData{MeshPrimitive::Triangles,
+        std::move(indexData), MeshIndexData{indices},
+        std::move(vertexData), std::move(attributeData),
+        UnsignedInt(indexCount)};
+
+    const bool generateIndices = configuration().value<bool>("generateIndices");
+    if (generateIndices)
+        meshData = MeshTools::removeDuplicates(meshData);
+
+    /* GCC 4.8 needs extra help here */
+    return Containers::optional(std::move(meshData));
+}
+
+UnsignedInt UfbxImporter::doMaterialCount() const {
+    return UnsignedInt(_state->scene->materials.count);
+}
+
+Int UfbxImporter::doMaterialForName(Containers::StringView name) {
+    return typedId(ufbx_find_element_len(_state->scene.get(), UFBX_ELEMENT_MATERIAL, name.data(), name.size()));
+}
+
+Containers::String UfbxImporter::doMaterialName(UnsignedInt id) {
+    return _state->scene->materials[id]->name;
+}
+
+Containers::Optional<MaterialData> UfbxImporter::doMaterial(UnsignedInt id) {
+    ufbx_material *material = _state->scene->materials[id];
+
+    struct MaterialMappingList {
+        Containers::ArrayView<const MaterialMapping> mappings;
+        Containers::ArrayView<ufbx_material_map> maps;
+    };
+
+    MaterialMappingList mappingLists[] = {
+        { Containers::arrayView(materialMapsFbx), Containers::arrayView(material->fbx.maps) },
+        { Containers::arrayView(materialMapsPbr), Containers::arrayView(material->pbr.maps) },
+    };
+
+    Containers::BitArray addedAttributes { ValueInit, 128 };
+
+    UnsignedInt maxAttributes = 0;
+    for (MaterialMappingList &list : mappingLists) {
+        /* Each mapping may generate multiple mappings, estimate for wrost case
+           Value + Texture + TextureMatrix + TextureCoordinates */
+        maxAttributes += list.mappings.size() * 4;
+
+        #ifndef CORRADE_NO_ASSERT
+        for (const MaterialMapping &mapping : list.mappings) {
+            CORRADE_INTERNAL_ASSERT(UnsignedInt(mapping.attrib) < addedAttributes.size());
+            CORRADE_INTERNAL_ASSERT(UnsignedInt(mapping.texture) < addedAttributes.size());
+            CORRADE_INTERNAL_ASSERT(UnsignedInt(mapping.textureCoordinates) < addedAttributes.size());
+            CORRADE_INTERNAL_ASSERT(UnsignedInt(mapping.textureMatrix) < addedAttributes.size());
```

Comment:
Is the `addedAttributes.size()` (128) queryable from the ufbx header in some way, or at least `static_assert()`able? For example

```cpp
constexpr std::size_t MaxAttributeIndex = 128;
static_assert(MaxAttributeIndex >= UFBX_MATERIAL_FBX_MAP_COUNT, &&
              MaxAttributeIndex >= UFBX_MATERIAL_PBR_MAP_COUNT, "ow!");
Containers::BitArray addedAttributes{ValueInit, MaxAttributeIndex};
```

In which case this inner loop wouldn't need to be here, and neither really the outer loop. The growable arrays have an ~exponential growth ratio, so I'm thinking it could be simpler to just rely on them doing sane reallocations instead of complex size estimates.

Reply:
The attributes are indexed with `MaterialAttribute` as index rather than ufbx ones, so I would need to know the maximum value for "standard" `MaterialAttribute` values for that. However for the issue below I'll probably rewrite this bit now so that's another way to solve this :D

Reply:
Ah, right, I looked at it the wrong way, sorry. At the moment there's no way to get the `MaterialAttribute` max value.